src/broadcast/scrobblingmanager.cpp

```diff
@@ -0,0 +1,210 @@
+{
+    m_CPGuiTick.connectValueChanged(SLOT(slotGuiTick(double)));
+    connect(&PlayerInfo::instance(),SIGNAL(currentPlayingTrackChanged(TrackPointer)),
+            &m_broadcaster,SLOT(slotNowListening(TrackPointer)));
+    m_broadcaster
+        .addNewScrobblingService(new FileListener("nowListening.txt"));
+    startTimer(1000);
+}
+
+
+void ScrobblingManager::slotTrackPaused(TrackPointer pPausedTrack) {         
+    QMutexLocker locker(&m_mutex);
+    bool allPaused = true;
+    TrackInfo *pausedTrackInfo = nullptr;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pPausedTrack) {
+            pausedTrackInfo = trackInfo;
+            for (QString playerGroup : trackInfo->m_players) {
+                BaseTrackPlayer *player = m_pManager->getPlayer(playerGroup);
+                if (!player->isTrackPaused())
+                    allPaused = false;
+            }
+            break;
+        }                   
+    }
+    if (allPaused && pausedTrackInfo)
+        pausedTrackInfo->m_trackInfo.pausePlayedTime();        
+}
+
+void ScrobblingManager::slotTrackResumed(TrackPointer pResumedTrack) {
+    BaseTrackPlayer *player = qobject_cast<Deck*>(sender());
+    DEBUG_ASSERT(player);    
+    if (player == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received resumed signal.";
+        return;
+    }    
+    if (isTrackAudible(pResumedTrack,player)) {        
+        QMutexLocker locker(&m_mutex);
+        for (TrackInfo *trackInfo : m_trackList) {
+            if (trackInfo->m_pTrack == pResumedTrack) {
+                trackInfo->m_trackInfo.resumePlayedTime();
+                break;
+            }
+        }  
+    }
+}
+
+void ScrobblingManager::slotLoadingTrack(TrackPointer pNewTrack, TrackPointer pOldTrack) {
+    BaseTrackPlayer *sourcePlayer = 
+        qobject_cast<BaseTrackPlayer*>(sender());
+    DEBUG_ASSERT(sourcePlayer);
+    if (pOldTrack) {
+        m_tracksToBeReset.append(TrackToBeReset(pOldTrack,
+            sourcePlayer->getGroup()));
+    }
+}
+
+void ScrobblingManager::slotNewTrackLoaded(TrackPointer pNewTrack) {
+    //Empty player gives a null pointer.
+    if (!pNewTrack)
+        return;          
+    BaseTrackPlayer *player = qobject_cast<BaseTrackPlayer*>(sender());    
+    DEBUG_ASSERT(player);
+    QMutexLocker locker(&m_mutex);    
+    bool trackAlreadyAdded = false;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pNewTrack) {
+            trackInfo->m_players.append(player->getGroup());               
+            trackAlreadyAdded = true;
+            break;
+        }
+    }
+    if (!trackAlreadyAdded) {
+        TrackInfo *newTrackInfo = new TrackInfo(pNewTrack);
+        newTrackInfo->m_players.append(player->getGroup());
+        m_trackList.append(newTrackInfo);                
+        connect(&m_trackList.last()->m_trackInfo,SIGNAL(readyToBeScrobbled(TrackPointer)),
```

Comment:
Why do you establish a signal/slot connection between all instances of the internal TrackTimingInfo class and the broadcaster instead of only using a single connection between the ScrobblingManager and the broadcaster? You send the affected track with as the payload data with every signal and the track is bound to each instance of TrackTimingInfo anyway. If you need more context than just the TrackPointer then you could always collect this in a dedicated signal data class.

Reply:
I don't see how it is any different, you're just adding one level of indirection. In the end, the broadcaster manages the info on the grace period timer, the scrobbling manager doesn't care about it. Could you please explain why you think this is a bad idea?

Reply:
Simple and robust design? It doesn't make sense to use n connections when 1 is sufficient and all the receiver needs to know about the context is already part of the message.

The track info objects are an implementation detail of your class that should not be visible from the outside. The connections make them visible, indirectly. You are also not allowed to delete those objects like you do now while some signals might still be pending. You have to delete them with QObject::deleteLater(). Otherwise the receiver will get a dangling sender pointer. But all this doesn't matter if you lift the connection up to the next level.

Reply:
Correct me if I'm wrong but if they live on the same thread aren't they under a direct connection? Then there are no pending signals because the slot is called immediately. However, I didn't think about exposing implementation details so you're right, I will call them from the scrobbling manager.